src/grift/type-check.rkt

```diff
@@ -419,127 +649,221 @@ The type rules for core forms that have interesting type rules
+                   (gvector-length-type-rule t))]
+          [(Create-tuple (app map-recur e* t*))
+           (values (Create-tuple e*) (tuple-type-rule t*))]
+          [(Tuple-proj (app recur e t) i)
+           (values (Tuple-proj e i) (tuple-proj-type-rule t i))]
+          [other (error 'type-check "unmatched ~a" other)]))
+      )
+    (values (Ann new-exp (cons src type)) type))
+  (recur exp))
 
 
 ;; Type checks the rhs to be consistent with type annotation if
 ;; provided the resulting type is the type of the annotation.
-(: mk-tc-binding :
-   Src (-> S0-Expr (values S1-Expr Grift-Type)) -> (-> S0-Bnd S1-Bnd))
+(: tc-binding :
+   Src (-> G0-Ann-Expr (values G1-Ann-Expr Grift-Type)) -> (-> G0-Bnd G1-Bnd))
 (define ((tc-binding src tc-expr) bnd)
-  (match-define (Bnd id id-type rhs) bnd)
+  (match-define (Bnd id t0 rhs) bnd)
   (define-values (new-rhs rhs-type) (tc-expr rhs))
-  (Bnd id (let-binding-type-rule id-type rhs-type id src) new-rhs))
-
-
-
+  (define t1 (let-binding-type-rule (and t0 (validate-type src t0)) rhs-type id src))
+  (Bnd id t1 new-rhs))
+
+(: map-tc-binding :
+   Src (-> G0-Ann-Expr (values G1-Ann-Expr Grift-Type))
+   -> (-> G0-Bnd* G1-Bnd*))
+(define ((map-tc-binding src tc-expr) bnd*)
+  (map (tc-binding src tc-expr) bnd*))
+
+
+(: map-switch-case2 :
+   (All (A B C)
+        (A -> (Values B C))
+        (Switch-Case A) ->
+        (Values (Switch-Case B) C)))
+(define (map-switch-case2 f c)
+  (define-values (r0 r1) (f (cdr c)))
+  (values (cons (car c) r0) r1))
+(: map-switch-case2* :
+   (All (A B C)
+        (A -> (Values B C))
+        (Switch-Case* A)
+        ->
+        (Values (Switch-Case* B) (Listof C))))
+(define (map-switch-case2* f a*)
+  (for/lists ([b* : (Switch-Case* B)]
+              [c* : (Listof C)])
+             ([a a*])
+    (map-switch-case2 f a)))
+
+
+;; resolve :: takes a type and converts it to a equivalent type
+;; where the outermost type is a concrete type constructor (like ->, Tuple, etc.) as
+;; opposed to logical (like Rec).
+;; NOTE: this should only be called on types that have been converted to normal form. 
+(define (resolve t)
+  (let rec ([t t])
+    (match t
+      [(Mu s) (resolve (grift-type-instantiate s t))]
+      [t t])))
+
+(define (type->sexp t [e : (hash)])
+  (let rec ([t t])
+    (match t 
+      [(FVar id) (hash-ref e id)]
+      [(Fn n a* r) `(,@(map rec a*) -> ,(rec r))]
+      [(Mu s)
+       (define u (next-uid! ""))
+       (define X (gensym "X"))
+       (define t0 (grift-type-instantiate s (FVar u)))
+       `(Rec ,X ,(type->sexp t0 (hash-set e u X)))]
+      [(STuple n a*) `(Tuple ,@(map rec a*))]
+      [(GVect a) `(GVect ,(rec a))]
+      [(GRef a) `(GRef ,(rec a))]
+      [(MVect a) `(MVect ,(rec a))]
+      [(MRef a)  `(MRef ,(rec a))]
+      [(Unit) '()]
+      [other
+       (define-values (nfo _0) (struct-info other))
+       (define-values (sym _1 _2 _3 _4 _5 _6 _7) (struct-type-info nfo))
+       sym])))
+
+(define (type->string t)
```

Comment:
This utility can be useful for debugging, do you think we should pull it out into a module for debugging utilities?

Reply:
Debugging or pretty printing... this could also be useful for user facing error message printing. Is this the start of a pretty printing library for ASTs?

Reply:
Sounds great!